### PR TITLE
[python-package] [ci] switch to PEP 517 / 518 builds (remove `setup.py`) (fixes #5061)

### LINKED ISSUE #5061: [RFC] [python] 'setup.py install' is deprecated
## Description

Running `python setup.py install` for the Python package raises the following deprecation warning:

> /Users/jlamb/mambaforge/lib/python3.9/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
  warnings.warn(

I found a very detailed write-up explaining this issue: ["Why you shouldn't invoke setup.py directly" (October 2021)](https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html).

Here are some of the main things we need to know:

> ...as of the last few years **all direct invocations of setup.py are effectively deprecated** in favor of invocations via purpose-built and/or standards-based CLI tools like [pip](https://pip.pypa.io/en/stable/), [build](https://pypa-build.readthedocs.io/en/stable/) and [tox](https://tox.wiki/en/latest/).

> As mentioned in [this issue](https://github.com/pypa/setuptools/issues/2088), direct setup.py invocations have effectively been unmaintained for several years now. Whenever someone raises an issue with a reproducer that involves invoking setup.py, the maintainers ask for a reproducer that doesn't hit this code path, and if one can't be found the issue is closed. Put another way, direct invocations of setup.py that currently work for you do so essentially by chance — if something breaks, you are on your own.

## Reproducible example

I observed this warning building from source on macOS 12.2.1, with Python v3.9.7 and `setuptools==59.8.0`.

```shell
git clone --recursive https://github.com/microsoft/LightGBM.git
cd LightGBM/python-package
```

This warning is raised with `python setup.py install`.

```shell
python setup.py install
```

It is also raised with `pip`.

```shell
pip install -v .
```

You can also see this warning on LightGBM's CI builds that use `python setup.py {command}`. For example, the most recent `Linux regular` job running on Azure DevOps ([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=12350&view=logs&j=c28dceab-947a-5848-c21f-eef3695e5f11&t=fa158246-17e2-53d4-5936-86070edbaacf)).

Screen Shot 2022-03-08 at 10 59 33 PM

## References

This warning was added in `setuptools=v58.3.0` (October 22, 2021), https://github.com/pypa/setuptools/pull/2824.

That PR mentions https://www.python.org/dev/peps/pep-0517/ as a description of the standard packages should meet to continue to be compliant with the main Python package management tools.

The warning comes from `setuptools.command.install.initialize_options()`: https://github.com/pypa/setuptools/blob/391bb5d4d09c9eb8d6b2b98968e623455ae0a384/setuptools/command/install.py#L32-L38.

`lightgbm` currently sub-classes `setuptools.command.install` to customize the behavior of `setup.py install`.

https://github.com/microsoft/LightGBM/blob/f6d654b737979738a443899c1be3edd06f9b7d29/python-package/setup.py#L220-L225

It also does that to customize `setup.py {install_lib, bdist_wheel, sdist}`.

https://github.com/microsoft/LightGBM/blob/f6d654b737979738a443899c1be3edd06f9b7d29/python-package/setup.py#L361-L366

## Why does this matter?

If nothing is changed, the LightGBM Python package might eventually not be installable with some future versions of Python / `setuptools`.

## What should be done?

I don't have a recommendation yet.

https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html, https://www.python.org/dev/peps/pep-0517/, and the other things they link to are very detailed and I'm not deeply familiar with the details of how `setuptools` works.

Just opening this up to track discussion about this issue.

### PR DESCRIPTION
**Pull request by jameslamb**
Fixes #5061.

This PR proposes removing the project's `setup.py`, and instead using `scikit-build-core` (https://github.com/scikit-build/scikit-build-core) as a build backend.

`scikit-build-core` has been developed with support from an NSF grant ([link](https://www.nsf.gov/awardsearch/showAward?AWD_ID=2209877&HistoricalAwards=false)) and is being developed by the same team developing `scikit-build`, which is used by other popular stats and machine learning projects, like:

* `cudf` (https://github.com/rapidsai/cudf/pull/10919)
* `cuml` (https://github.com/rapidsai/cuml/pull/4818)
* `opencv` ([code link](https://github.com/opencv/opencv-python/blob/474a1cc0ebf2086c596b60c050a9e1af658ff380/pyproject.toml#L4))
* `vaex` (https://github.com/vaexio/vaex/pull/1847)

*(I got that list from https://github.com/orgs/scikit-build/discussions/976)*

`scikit-build-core` is purpose-built for Python packages that use `CMake` to compile one or more other artifacts (in our case, the `lib_lightgbm` shared library). 

### What would happen if we did nothing?

Users would have to pin to older versions of `setuptools` and `pip` to be able to install `lightgbm` (which will make `lightgbm` harder to use alongside other packages). We've already started seeing evidence of that pain:

* #5897
* https://github.com/microsoft/LightGBM/issues/5061#issuecomment-1274724145

### Benefits of this approach

* users can use the newest versions of `pip`, `setuptools`, and `build` in environments where they install `lightgbm`
* no more need to maintain 400ish lines of Python code in `setup.py`
* makes it easier to compile wheels for more platforms in the future (e.g. MUSL-based wheels for Alpine Linux, `macosx-arm64` for the M1/M2 Macs)

### What other approaches could we take?

* could use `hatchling` like `xgboost` has recently done (see https://github.com/dmlc/xgboost/pull/9021#discussion_r1173171975)
    - *this has the advantage of supporting most of what `lightgbm` packaging currently supports, at the expense of needing to still maintain significant Python glue code*
* could use `scikit-build` (`scikit-build-core`'s predecessor)
    - *I don't think we should do this, as https://github.com/scikit-build/scikit-build says refers to `scikit-build-core` as "The next generation of scikit-build," and the author of both tools referred to `scikit-build` as "scikit-build classic" in https://github.com/orgs/scikit-build/discussions/815... implying that `scikit-build-core` is where most active development is going*
* could stop shipping source distributions entirely and use just a shell script to build wheels
    - *`catboost` was doing this until very recently*
* could switch build toolchains entirely, away from CMake to something like `meson`, `autotools`, `bazel`, etc. and the corresponding options for Python build backends
    - `numpy`, for example, is in the process of switching to `meson-python`: https://github.com/numpy/numpy/pull/23838

These are all possible routes to resolve #5061.

I'm proposing using `scikit-build-core` because it's powerful while also significantly simplifying the code that we have to manage in this project, and because it's maintained by members of PyPA who will keep it up to date with the latest changes in Python packaging (i.e. the latest accepted PEPs related to packaging).

### Existing features preserved with this approach

<details><summary>list of features (click me)</summary>

All of the follow commands are currently supported, and we should try to preserve as many of them as possible.

* `{command}` is `python setup.py` right now
* `{dist}` is either the name `lightgbm` (pull from PyPI) or a path to a `*.tar.gz` or `*.whl`

Installing from source tree.

```shell
{command} install
{command} install --precompile
{command} install --precompile --user
{command} install --mingw
{command} install --mpi
{command} install --nomp
{command} install --gpu
{command} install --gpu --opencl-include-dir=/usr/local/cuda/include/
{command} install --cuda
{command} install --hdfs
{command} install --bit32
```

Building an sdist.

```shell
{command} sdist
{command} sdist --formats gztar
```

Building wheels.

```shell
{command} bdist_wheel --plat-name=macosx --python-tag py3
{command} bdist_wheel --integrated-opencl --plat-name=$PLATFORM --python-tag py3
{command} bdist_wheel --integrated-opencl --plat-name=win-amd64 --python-tag py3
{command} bdist_wheel --gpu
{command} bdist_wheel --cuda
{command} bdist_wheel --mpi
```

Customized installation of an sdist (possibly from PyPI).

```shell
pip install {dist} --install-option=--nomp
pip install {dist} --install-option=--mpi
pip install {dist} --install-option=--gpu
pip install {dist} --install-option=--gpu --install-option="--opencl-include-dir=/usr/local/cuda/include/" --install-option="--opencl-library=/usr/local/cuda/lib64/libOpenCL.so"
pip install {dist} --install-option=--cuda
pip install {dist} --install-option=--hdfs
pip install {dist} --install-option=--mingw
pip install {dist} --install-option=--bit32
pip install {dist} --install-option=--time-costs
pip install {dist}[dask]
pip install --no-binary `:all:` {dist}
pip install --user {dist}
```

</details>

### Existing features lost with this approach

* automatically redirecting CMake logs to a file ([code link](https://github.com/microsoft/LightGBM/blob/f975d3fafcdbb3739dbe4eac40dc2b7e1e3244d7/python-package/setup.py#L95-L96))
    - *CMake logs will now just be written to stdout directly, and users will have to redirect it themselves if they want to save it to a file*
    - *honestly, I think this is an improvement... no more seeing a remote build like a CI job fail without knowing the underlying CMake error*
* using `MSBuild` to compile with `pip install {sdist}`, then multiple versions of MSVC ([link](https://github.com/microsoft/LightGBM/blob/f975d3fafcdbb3739dbe4eac40dc2b7e1e3244d7/python-package/setup.py#L168-L204))
    - *looks like `scikit-build` will check for multiple versions of MSVC:* https://github.com/scikit-build/scikit-build/blob/master/docs/generators.rst
    - *as of this PR, the only way to build the Python package with `MSBuild` will be to clone the repo and invoke `MSBuild.exe` directly*
* compiling the 32-bit version with `pip install {sdist}` ([link](https://github.com/microsoft/LightGBM/blob/f975d3fafcdbb3739dbe4eac40dc2b7e1e3244d7/python-package/setup.py#L250-L251))
    - *as of this PR, the only way to do that type of build will be to clone the repo and run `sh build-python.sh install --bit32`*

### CODE REVIEW

File: .ci/test_windows.ps1 [hunk 1]
```diff
@@ -96,8 +96,8 @@ elseif ($env:TASK -eq "bdist") {
   cd $env:BUILD_SOURCESDIRECTORY
   sh "build-python.sh" bdist_wheel --integrated-opencl ; Check-Output $?
   sh $env:BUILD_SOURCESDIRECTORY/.ci/check_python_dists.sh $env:BUILD_SOURCESDIRECTORY/dist ; Check-Output $?
-  cd dist; pip install --user @(Get-ChildItem *.whl) ; Check-Output $?
-  cp @(Get-ChildItem *.whl) $env:BUILD_ARTIFACTSTAGINGDIRECTORY
+  cd dist; pip install --user @(Get-ChildItem *py3-none-win_amd64.whl) ; Check-Output $?
+  cp @(Get-ChildItem *py3-none-win_amd64.whl) $env:BUILD_ARTIFACTSTAGINGDIRECTORY
```
Comment (jameslamb) [inline .ci/test_windows.ps1:100]: This ensures we get a loud error if the wheels are not tagged in the way we'd expect.

Since as far as I can tell, you can't pass a `--plat-name` through `python -m build` the way you could with  `python setup.py bdist_wheel --plat-name=win-amd64`.

To be clear... `scikit-build-core` is doing exactly the right thing here and producing the expected tags. This is just a precaution to help us detect if that changes, before publishing mistagged wheels to PyPI (e.g. like #4788).

File: python-package/README.rst [hunk 2]
```diff
@@ -84,7 +84,7 @@ Build GPU Version
 
 .. code:: sh
 
-    pip install lightgbm --install-option=--gpu
+    pip install lightgbm --config-setting=cmake.define.USE_GPU=ON
```
Comment (hcho3) [inline python-package/README.rst:87]: Pip exposes `--config-settings` option: https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-C. Is `--config-setting` coming from somewhere else?

Comment (jameslamb) [reply python-package/README.rst]: Oh wow, thank you! No you're absolutely right, this is a mistake on my part.

I was mixing two things... `pip` has `--config-settings` (with an `s`), and `build` has `--config-setting` (no `s`) ([code link](https://github.com/pypa/build/blob/e0c59cc4e17e63cfe9dbd4f366fae6d51b073282/src/build/__main__.py#L331-L339)).

I was following from https://github.com/scikit-build/scikit-build-core/blob/a145985892e4c904f4e4f089f987b4fef491543d/README.md?plain=1#L112-L113 and didn't read closely enough.

Just pushed https://github.com/microsoft/LightGBM/pull/5759/commits/41badff4121b6003c95e8dbc8e967348e1820e96 fixing that.

File: build-python.sh [hunk 3]
```diff
@@ -94,75 +94,79 @@ while [ $# -gt 0 ]; do
             then shift;
         fi
         BOOST_DIR="${1#*=}"
-        BUILD_ARGS="${BUILD_ARGS} --boost-dir='${BOOST_DIR}'"
+        BUILD_ARGS="${BUILD_ARGS} --config-setting=cmake.define.Boost_DIR='${BOOST_DIR}'"
```
Comment (guolinke) [inline build-python.sh:97]: ```--config-settings``` or ```--config-setting``` in this file?

Comment (jameslamb) [reply build-python.sh:97]: `--config-settings` when passed to `pip`, `--config-setting` when passed to `build` 😅 

https://github.com/microsoft/LightGBM/pull/5759#discussion_r1224471802

**Review by guolinke — APPROVED** [review]

**Review by hcho3 — APPROVED** [review]
I find it interesting that you were able to use `scikit-learn-core` without adding any custom glue code. I suppose that `build-python.sh` does much of the heavy-lifting of packaging the right files in the right place?

For now, XGBoost adds a custom PEP 517 glue code to marshal C++ source files, but in the future I may consider writing a bash script instead, should we decide to transition to `scikit-build-core`. Thanks for the inspiration!

**Comment by jameslamb** [comment]
Thanks so much for the review help @hcho3 ! I know you're busy enough with the other projects you work on, I always really appreciate your help here.

> build-python.sh does much of the heavy-lifting of packaging the right files in the right place

Yep, exactly! I described my perspective on this in #5837 if you want to see a bit more detail on how I'm thinking about it.

Essentially, I think any Python user of LightGBM whose needs aren't met by `pip install` (from PyPI) or `conda install` (from conda-forge) will already need to clone the repo and run some command(s). Making that "run a shell script" should be just as easy for those people as `python setup.py install` or `pip install .`, so it's worth switching to that in exchange for simplifying the build system for the artifacts we distribute via PyPI.

I'm also hoping that having a zero-glue-code setup with `scikit-build-core` will make it easier in the future to build wheels for more platforms, further reducing the need for people to build custom from the sources here in GitHub.

We've been used a similar approach with the R package in this repo for a few years, and I think it's worked well so far: https://github.com/microsoft/LightGBM/blob/master/build-cran-package.sh.

File: python-package/pyproject.toml [hunk 4]
```diff
@@ -1,3 +1,63 @@
...
+    "Development Status :: 5 - Production/Stable",
+    "Intended Audience :: Science/Research",
+    "License :: OSI Approved :: MIT License",
+    "Natural Language :: English",
+    "Operating System :: MacOS",
+    "Operating System :: Microsoft :: Windows",
+    "Operating System :: POSIX",
+    "Operating System :: Unix",
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3.7",
```
Comment (jmoralez) [inline python-package/pyproject.toml:13]: Shouldn't Python 3.6 be here given that we have `requires-python = ">=3.6"`?

Comment (jameslamb) [reply python-package/pyproject.toml:13]: oh sure, I can add that.

It's not there in `setup.py` and I was just copying from that.

https://github.com/microsoft/LightGBM/blob/a323c7583ccdffb7aff49e7f9ca72042aca18395/python-package/setup.py#L341-L355

**Review by jmoralez — APPROVED** [review]

File: python-package/pyproject.toml (hunk 4 above)
Comment (guolinke) [reply python-package/pyproject.toml:13]: refer to https://github.com/microsoft/LightGBM/pull/4891

Comment (jameslamb) [reply python-package/pyproject.toml:13]: Ah right, thanks @guolinke for reminding me ... in #4891 we dropped support, meaning:

* dropped that trove classifier `'Programming Language :: Python :: 3.6'` from the PyPI labels
* do not have any CI jobs testing Python 3.6

And we decided not to move up to `requires-python = ">=3.7"` so that people could *technically* try to build for Python 3.6 if they needed that a bit longer (as @jmoralez also referred to in https://github.com/microsoft/LightGBM/pull/5765#issuecomment-1454861675).

By the way, I tried tonight and it looks like `lightgbm` is still at least minimally compatible with Python 3.6 🎉 .

```shell
docker run \
    --rm \
    -it python:3.6 \
    bash

apt-get update
apt-get install \
    --no-install-recommends \
    -y \
        cmake

git clone --recursive \
    https://github.com/microsoft/LightGBM.git

cd LightGBM
sh ./build-python.sh install --nomp

pip install pandas
python ./examples/python-guide/simple_example.py
```

logs:

```text
Loading data...
Starting training...
[LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.002058 seconds.
You can set `force_col_wise=true` to remove the overhead.
Training until validation scores don't improve for 5 rounds
Did not meet early stopping. Best iteration is:
[20]	valid_0's l2: 0.198063	valid_0's l1: 0.431129
Saving model...
Starting predicting...
The RMSE of prediction is: 0.4450426449744025
```

I don't feel strongly either way about whether or not to have the Python 3.6 trove classifier. Since @StrikerRUS removed it in #4891, I think we should defer to his judgment and continue to omit it. Is that ok @jmoralez ? If, after reading this, you still feel strongly that it should be included since `lightgbm` at least *minimally* supports Python 3.6, I'll add it.

Comment (jmoralez) [reply python-package/pyproject.toml:13]: That's ok, I agree with not adding it back.

Comment (jameslamb) [reply python-package/pyproject.toml:13]: Alright, thanks for talking through it with us! In another PR, I might put up a minimal CI job that runs exactly that dockerized little script I just shared above. So we can rely on CI to catch issues of the form "this PR would be a hard compatibility break because it's not even syntactically valid Python 3.6".

I don't want to take on any more support than that for Python 3.6, but would be useful (I think) to get automated feedback at that level at least, and I'd expect that job shouldn't require too much maintenance burden.

Comment (jameslamb) [reply python-package/pyproject.toml:13]: For those finding this conversation from search... here's the PR I promised: #5936.
